File: DP/maxSubArray.py
class Solution:
    def maxSubArray(self, nums):
        """动态规划
        假设对于元素i，所有以它前面的元素结尾的子数组的长度都已经求得，那么以第i个元素结尾且和最大的连续子数组实际上         ，要么是以第i-1个元素结尾且和最大的连续子数组加上这个元素，要么是只包含第i个元素，
        即sum[i] = max(sum[i-1] + a[i], a[i])
        """
        """
        :type nums: List[int]
        :rtype: int
        """
        globalMax = nums[0]
        localMax = nums[0]
        for item in nums[1:]:
            if localMax > 0:
                localMax += item
            else:
                localMax = item
            if localMax>globalMax:
                globalMax = localMax

        return globalMax



if __name__ == "__main__":
    s = Solution()
    nums = [-2,1,-3,4,-1,2,1,-5,4]
    print(s.maxSubArray1(nums))

File: DP/maxSubArray.c
#include <stdio.h>

int max(int x, int y){
    int result;
    if (x > y){
        result = x;
    }else{
        result = y;
    }
    
    return result;
}


int maxSubArray(int* nums, int numsSize){
    if (numsSize == 0){
        return 0;
    }

    int local_max_sum = nums[0];
    int global_max_sum = local_max_sum;

    for (int i=1;i<numsSize;i++){
        local_max_sum = nums[i]+local_max_sum>nums[i]?nums[i]+local_max_sum:nums[i];

        if(local_max_sum>global_max_sum){
            global_max_sum = local_max_sum;
        }
        printf("%d\n", local_max_sum);
    }
    
    return global_max_sum;
}

int main(){
    int nums[9] = {-2,1,-3,4,-1,2,1,-5,4};
    int numsSize = 9;
    int maxSum;

    maxSum = maxSubArray(nums, numsSize);
    printf("%d", maxSum);
}